### ISSUE TITLE
Some problem about DDS_RELIABILITY_BEST_EFFORT(QoS)

### ISSUE THREAD

**Issue by jwcesign** [eclipse-cyclonedds/cyclonedds#32]
Hi,
Recently, I want to know the throughput of dds without losing samples.  No matter what parameter I set(at the same host), it always loses samples when I run the Throughput example with DDS_RELIABILITY_BEST_EFFORT QoS setting.
~~~
# terminal 1
./ThroughputSubscriber
Cycles: 0 | PollingDelay: 0 | Partition: Throughput example
=== [Subscriber] Waiting for samples...
=== [Subscriber] Payload size: 4000 | Total received: 2 samples, 8016 bytes | Out of order: 0 samples Transfer rate: 0.00 samples/s, 0.00 Mbit/s
=== [Subscriber] Payload size: 4000 | Total received: 503 samples, 2016024 bytes | Out of order: 0 samples Transfer rate: 497.05 samples/s, 14.94 Mbit/s
=== [Subscriber] Payload size: 4000 | Total received: 1004 samples, 4024032 bytes | Out of order: 0 samples Transfer rate: 496.97 samples/s, 14.94 Mbit/s
=== [Subscriber] Payload size: 4000 | Total received: 1504 samples, 6028032 bytes | Out of order: 0 samples Transfer rate: 496.07 samples/s, 14.94 Mbit/s
=== [Subscriber] Payload size: 4000 | Total received: 2002 samples, 8024016 bytes | Out of order: 0 samples Transfer rate: 494.16 samples/s, 14.94 Mbit/s
^C=== [Subscriber] Payload size: 4000 | Total received: 2490 samples, 9979920 bytes | Out of order: 0 samples Transfer rate: 493.90 samples/s, 15.24 Mbit/s

Total received: 2490 samples, 9979920 bytes
Out of order: 0 samples
Average transfer rate: 622.50 samples/s, 19.75 Mbit/s
------------------------------------------------------------------------------
# terminal 2
./ThroughputPublisher 4000 20 10 5
payloadSize: 4000 bytes burstInterval: 20 ms burstSize: 10 timeOut: 5 seconds partitionName: Throughput example

=== [Publisher]  Waiting for a reader ...
=== [Publisher]  Writing samples...
=== [Publisher]  Timed out, 2500 samples written.
~~~
In my test, the pub writes 2500 samples but the sub just gets 2490 samples. In my opinion, if the load is lite and the pub/sub are running at same host, it should not lose samples. why does this happen?

**Comment by eboasson** [comment]
Hi,

You are entirely correct that with ~40kB every 20ms there should no packet loss over the loopback interface, and that, from some point onward, everything is received is also indicated by the "out of order" count remaining at 0.

So if I were to hazard a guess, it would be that typically the first batch of samples doesn't make it. Even though the publisher waits until it has discovered the reader, that in itself does not mean the reader has discovered the writer. Both need to have happened for the data to be delivered.

For best-effort data — "unreliable" it used to be called when I started working, a far more honest term — there is not much you can do about it. However, even for volatile, reliable data, this can be an issue. In that case, the writer will store the data in its WHC and advertise its existence in the Heartbeat messages, but the reader will ignore all this data as it really is historical data from its perspective. While arguably correct, it is counter-intuitive, and that always is a bad thing if the intuitive behaviour would also be correct.

If instead the reader were to simply request everything that's advertised, and for some reason the writer has a lot of data in its history (it could be a transient-local writer, but it is also possible for volatile writers if there are other readers around), the behaviour would not only be counter-intuitive, but also incorrect.

I am pretty sure I know a way that's within the allowed behaviour of the DDSI specification, and that is for a writer to send a Heartbeat specifically to a newly matched reader, indicating from what the first sequence number is that the writer has written, or will write, since discovering that reader. The information is available internally, the only problem is communicating it. But it seems to me that if that Heartbeat simply avoids advertises anything published later, it would be conforming and the reader would be able to use this information to start delivering samples from the sequence number one would expect it to. The behaviour would then perhaps have to treat Cyclone slightly differently from the other DDS implementations, but that's definitely feasible.

Long story short: can you verify that it is always the first samples that are missing?

P.S. There is no easy fix at application level. Of course one can build a protocol on top of DDS to ensure the matching has completed on both sides, but that's not a very sensible approach. In practice a short sleep will do, but in the long run, for reliable data it will have to be a fix at protocol level, perhaps the one I outlined, and for best-effort data, I don't think there's any solution.

P.P.S. Since one should only use best-effort when one can afford to lose data, it should not be an issue.

**Comment by jwcesign** [comment]
Nice answer. "Even though the publisher waits until it has discovered the reader, that in itself does not mean the reader has discovered the writer. Both need to have happened for the data to be delivered." I agree with this opinion.

**Comment by eboasson** [comment]
No further comments for a long time, so closing.
